### PR TITLE
Optimize finding free timestamp query slots in our Vulkan layer

### PR DESCRIPTION
**Pull request by ronaldfw**
We optimize finding a free timestamp query slot. The previous code ran
over a vector of all possible slots until a free one (based on the
vector entry) was found. This was problematic when many, and in
particular all, slots were occupied already, leading to massive game
slowdown and a lot of CPU time spent in finding the next free slot.

This is mostly noticeable when running out of query slots. With a
maximum of 10'000 slots, the previous code had ~84% of exclusive
callstack samples in TimerQueryPool::NextReadyQuerySlot. For the new
code, this number went down to ~0.14%.

For validation purposes we keep the vector of slot states and check
allocated slots and reset slots against the expected state.

Tested: Manually ran Vulkan layer on two test applications. Profiled
Orbit layer / game performance using Orbit.
Bug: http://b/181639638

### CODE REVIEW

File: src/OrbitVulkanLayer/TimerQueryPool.h [hunk 1]
```diff
@@ -51,7 +54,9 @@ class TimerQueryPool {
       std::vector<SlotState> slots{num_timer_query_slots_};
       std::fill(slots.begin(), slots.end(), SlotState::kReadyForQueryIssue);
       device_to_query_slots_[device] = slots;
-      device_to_potential_next_free_index_[device] = 0;
+      std::vector<uint32_t> free_slots(num_timer_query_slots_);
+      std::iota(free_slots.begin(), free_slots.end(), 0);
```
Comment (florian-kuebler) [inline src/OrbitVulkanLayer/TimerQueryPool.h:58]: Maybe add a comment, that at the beginning, every slot is considered empty (I needed to look up the iota method).

File: src/OrbitVulkanLayer/TimerQueryPool.h [hunk 2]
```diff
@@ -76,30 +81,24 @@ class TimerQueryPool {
...
-  // In case it successfully found a slot, the index will be written to the given `allocated_index`.
+  // Returns a free query slot from the device's pool if one still exists. It returns `false` if all
+  // slots are occupied and true otherwise. If successful, the index will be written to the given
+  // `allocated_index`.
   //
   // Note that the pool must be initialized using `InitializeTimerQueryPool` before.
   // See also `ResetQuerySlots` to make occupied slots available again.
   [[nodiscard]] bool NextReadyQuerySlot(VkDevice device, uint32_t* allocated_index) {
     absl::WriterMutexLock lock(&mutex_);
-    CHECK(device_to_potential_next_free_index_.contains(device));
```
Comment (florian-kuebler) [inline src/OrbitVulkanLayer/TimerQueryPool.h:87]: That should probably be a CHECK(device_to_free_slots_.contains(device));

File: src/OrbitVulkanLayer/TimerQueryPool.h [hunk 3]
```diff
@@ -76,30 +81,24 @@ class TimerQueryPool {
...
-
-    return false;
+    std::vector<uint32_t>& free_slots = device_to_free_slots_.at(device);
+    if (free_slots.empty()) {
+      return false;
+    }
+    *allocated_index = free_slots.back();
+    CHECK(device_to_query_slots_.at(device)[*allocated_index] == SlotState::kReadyForQueryIssue);
+    device_to_query_slots_.at(device)[*allocated_index] = SlotState::kQueryPendingOnGpu;
+    free_slots.pop_back();
```
Comment (florian-kuebler) [inline src/OrbitVulkanLayer/TimerQueryPool.h:100]: Nit/Personal preference: I would move the free_slots.pop_back up to the free_slots.back();

**Review by florian-kuebler — APPROVED** [review]
Thanks. LGTM!

File: src/OrbitVulkanLayer/TimerQueryPool.h (hunk 1 above)
Comment (ronaldfw) [reply src/OrbitVulkanLayer/TimerQueryPool.h:58]: Done.

File: src/OrbitVulkanLayer/TimerQueryPool.h (hunk 2 above)
Comment (ronaldfw) [reply src/OrbitVulkanLayer/TimerQueryPool.h:87]: Yes, good idea. Done.

File: src/OrbitVulkanLayer/TimerQueryPool.h (hunk 3 above)
Comment (ronaldfw) [reply src/OrbitVulkanLayer/TimerQueryPool.h:100]: Yes, agree. Done.

File: src/OrbitVulkanLayer/TimerQueryPool.h (hunk 1 above)
Comment (florian-kuebler) [reply src/OrbitVulkanLayer/TimerQueryPool.h:59]: Thanks!

File: src/OrbitVulkanLayer/TimerQueryPool.h (hunk 2 above)
Comment (florian-kuebler) [reply src/OrbitVulkanLayer/TimerQueryPool.h:87]: Thanks!

File: src/OrbitVulkanLayer/TimerQueryPool.h (hunk 3 above)
Comment (florian-kuebler) [reply src/OrbitVulkanLayer/TimerQueryPool.h]: Thanks!

**Comment by ronaldfw** [comment]
Thanks for the review, @florian-kuebler

**Pull request merged**